myia/composite.py

```diff
@@ -595,3 +597,79 @@ def _array_zero(xs):
     nonleaf=(Tuple, List, Class),
     fn_leaf=_leaf_zeros_like
 )
+
+
+@core
+def _cast_helper(x, model):
+    t = typeof(model)
+    if hastype(model, Array):
+        return scalar_to_array(scalar_cast(x, t.elements))
+    else:
+        return scalar_cast(x, t)
+
+
+class GradOperation(MetaGraph):
+    """Implements the grad(f) operation.
+
+    grad(f)(x, ...) returns df(x, ...)/dx. Derivatives of other inputs are
+    thrown out.
+
+    TODO: This currently will not work on primitives, but it is an easy fix.
+    We just need to know how many parameters f takes.
```

Comment:
I feel that it's easy enough to wrap a primitive in a graph if you really need it that we don't need to implement that.

Reply:
We'll see.